modules/db/src/main/scala/gem/dao/package.scala

```diff
@@ -49,8 +49,8 @@ package object dao extends MoreTupleOps with ToUserProgramRoleOps {
   implicit val InstantMeta: Meta[Instant] =
     Meta[Timestamp].nxmap(_.toInstant, Timestamp.from)
 
-  implicit val LocationMeta: Meta[Location] =
-    Meta[List[Int]].nxmap(Location.fromList, _.toList)
+  implicit val LocationMeta: Meta[Location.Middle] =
+    Meta[List[Int]].nxmap(Location.unsafeMiddle(_), _.toList)
```

Comment:
If you have a parameterized type `case class Foo[A](f: A => A)` for instance, where `A` appears in both covariant ("return value") and contravariant ("argument") position and you then wish to derive a `Foo[B]` from it, you need to provide both `g: B => A` and `h: A => B` from which you can construct a `B => B` as `g andThen f andThen h`. Functors of this type are called exponential or invariant functors, which is where `xmap` came from … in cats it's called `imap`.

In any case because `Meta` provides mapping to read *and* write a column value it's an invariant functor like `Foo` above.

The `nxmap` is just a convenience that allows you to ignore the `null` case, since JDBC speaks in terms of `null` and we don't. The next-higher abstraction `Atom` doesn't allow you to see `null`; from `Meta[A]` you get both `Atom[A]` and `Atom[Option[A]]` for the non-nullable and nullable cases, respectively, and user-level doobie code never observes `null`.

More generally you can see invariant functors as a supertype, where covariant and contravariant functors as specializations that ignore one or the other of the `xmap` arguments. There is a fourth type called phantom functors which ignore *both* arguments and are thus both covariant *and* contravariant, which would be the case for something like `case class Phantom[A](i: Int)` where `A` is never used so it can vary in either direction.